### PR TITLE
add archive address flag

### PR DESCRIPTION
**Pull request by zhangchiqing**
Close https://github.com/onflow/flow-go/issues/4186

### CODE REVIEW

File: engine/access/rpc/backend/backend_scripts.go [hunk 1]
```diff
@@ -81,6 +82,31 @@ func (b *backendScripts) ExecuteScriptAtBlockHeight(
 	return b.executeScriptOnExecutionNode(ctx, blockID, script, arguments)
 }
 
+func findScriptExecutors(
```
Comment (zhangchiqing) [inline engine/access/rpc/backend/backend_scripts.go:85]: This function is the main change, as well as the function that calls it.

It wraps the original `executionNodesForBlockID`, and check whether archiveAddress is configured, which will take the priority to be used as script executor.

File: cmd/access/node_builder/access_node_builder.go [hunk 2]
```diff
@@ -115,6 +115,7 @@ type AccessNodeConfig struct {
 	stateStreamConf              state_stream.Config
 	stateStreamFilterConf        map[string]int
 	ExecutionNodeAddress         string // deprecated
+	ArchiveNodeAddressList       []string
```
[the remark this answers is not in the document]
Comment (zhangchiqing) [reply cmd/access/node_builder/access_node_builder.go:118]: >executionNodesForBlockID route the script to the "correct" version of execution node

Can you point out the code for this "route" logic?

[the remark this answers is not in the document]
Comment (pattyshack) [reply cmd/access/node_builder/access_node_builder.go:118]: (I'm not familiar with this part of the stack) My question is how are request routed in general / where is the routing decision being made.

I know that we maintain execution nodes from each spork for historical query.  My question is:
1. Do all requests end up in the same set of access nodes, then gets forwarded to the "correct" spork-versioned execution node, or
2. Do we have maintain multiple sets of access nodes (one of each spork version), and something sit in front of the access node and makes that routing decision.

If the architecture resemble option 1. then my previous comment applies.  You need to add equivalent routing functionality for archive nodes.  If the architecture resemble option 2, then the PR lgtm.

[the remark this answers is not in the document]
Comment (zhangchiqing) [reply cmd/access/node_builder/access_node_builder.go:118]: Yes, it's the option 2. 

We keep one AN per spork, and [we publish the spork height for past sporks](https://developers.flow.com/nodes/node-operation/past-sporks), and the address of historical AN for querying state.

[the remark this answers is not in the document]
Comment (zhangchiqing) [reply cmd/access/node_builder/access_node_builder.go:118]: For transactions, the AN will query the historical ANs to find the transaction by ID
https://github.com/onflow/flow-go/blob/master/engine/access/rpc/backend/backend_transactions.go#L194

**Review by SaveTheRbtz — APPROVED** [review]

File: cmd/access/node_builder/access_node_builder.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (pattyshack) [reply cmd/access/node_builder/access_node_builder.go:118]: ok, your change make sense given this context.  thx

**Review by pattyshack — APPROVED** [review]

**Comment by zhangchiqing** [comment]
bors merge

**Review by koko1123 — APPROVED** [review]

**Pull request merged**